imgpkg/002-recursive-bundles/README.md

```diff
@@ -0,0 +1,304 @@
+    3. The Packager is responsible for organizing the configuration for all the Applications
+    4. The Packager creates a single bundle.
+
+    This approach is complicated to maintain in the long run for the Packager. In case an Application Team creates a new update on the application, this process would have to be repeated by the Packager. This process can become a bottleneck because the Packager would need to have a deep understanding of each application and the configuration needs.
+- Multiple bundles:
+    1. Each Application Team would generate their Bundle with their application
+    2. The Packager collects all the Bundles SHA
+    3. The Packager creates a document with all the Bundle SHA
+    4. The Packager creates overlay's and adds some utilities to use during installation
+    5. The Packager generates a Bundle with these artifacts
+    6. The Consumer retrieves the Bundle generated by the Packager
+    7. The Consumer retrieves every other Bundle that is required
+
+    This approach has challenges for both Packager and Consumer
+    The Consumer would have to download multiple separate bundles and follow the instructions from the Packager to install all the applications
+    The Packager would have to collect the information about the applications and generate an installation guide for the Consumer.
+
+### Mitigate Pain Points
+
+In this proposal, we adapt `imgpkg` to try to mitigate some of the current pains.
+Given that `imgpkg` already supports [bundles](https://carvel.dev/imgpkg/docs/latest/resources/#bundle) with a simple extension that allows bundles to contain other bundles can be created to help with the problem.
+
+The workflow that this change enables is the following:
+1. Each Application Team generates a bundle with all the needed images and the configuration
+2. The Packager would collect the applications Image Bundle SHA's for all applications
+3. The Packager can provide some overlay's for the configuration and add some utilities that are used during installation
+4. Create a Bundle that contains all application Image Bundles artifacts generated in the previous point
+5. The Application Consumer would be given a single Bundle Image that can be downloaded and consumed
+
+The benefits of this approach versus the current state are
+- Each Application Team manages their application's Bundle. The team that builds the application will manage all the requirements
+- The Packager just needs to collect the Bundle SHA for each application
+- The Consumer can get all the needed resources from a single Bundle to install all the applications needed
+
+# Proposed changes
+
+In this section, the proposal will only talk about parts that will be changed based on the Recursive Bundle solution
+
+## Pushing a bundle
+
+The major proposed change in this section is to remove the validation done today that checks if the bundle being pushed contains bundles in the `.imgpkg/images.yml` Lock file.
```

Comment:
Has there been any thought about keeping `images.yml` exclusive to images and using something like the [BundleLock](https://carvel.dev/imgpkg/docs/latest/resources/#bundlelock-configuration) in the case of nested bundles?

Reply:
I think this separation could be an interesting idea, nevertheless, having a single lock file with all the images feels more natural. This separation would help `imgpkg` implementation because now it knows which image is based on the file it picks it up from. But it would make the life of the bundle create harder because this separation manually would have to be done and maintained by hand.
Also, it would increase the complexity of the image validation because now we would have to ensure that the creator placed bundles in one Lock file and images in a different Lock File.

After going through these pros and cons I would prefer to keep all the images in a single Lock file, to give the user the perception that `imgpkg` handles images. These images could be bundles or regular images, but in the end, they are images.

Reply:
Yeah, I am not strongly opinionated on it and agree it feels like better UX to have a single file. I think it calls into question the need for a BundleLock file is more what I am getting at, but that's outside the scope of this.

Reply:
+1 on single file for all referenced assets.